pkg/os/shell/shell.go

```diff
@@ -151,3 +171,50 @@ func IsWindowsSubsystemLinux() bool {
 	}
 	return false
 }
+
+func (p *RealProcess) Parent() (AbstractProcess, error) {
+	parentProcess, err := p.Process.Parent()
+	if err != nil {
+		return nil, err
+	}
+	return &RealProcess{parentProcess}, nil
+}
+
+func createCurrentProcess() AbstractProcess {
+	currentProcess, err := process.NewProcess(cast.ToInt32(os.Getpid()))
+	if err != nil {
+		return nil
+	}
+	return &RealProcess{currentProcess}
+}
+
+// detectShellByCheckingProcessTree attempts to identify the shell being used by
+// examining the process tree starting from the given process ID. This function
+// traverses up to ProcessDepthLimit levels up the process hierarchy.
+// Parameters:
+//   - pid (int): The process ID to start checking from.
+//
+// Returns:
+//   - string: The name of the shell if found (e.g., "zsh", "bash", "fish");
+//     otherwise, an empty string is returned if no matching shell is detected
+//     or an error occurs during the process tree traversal.
+//
+// Examples:
+//
+//	shellName := detectShellByCheckingProcessTree(1234)
+func detectShellByCheckingProcessTree(p AbstractProcess) string {
+	for p != nil {
+		processName, err := p.Name()
+		if err != nil {
+			return ""
+		}
+		if crcos.IsPresentInList(supportedShell, processName) {
```

Comment:
This would be easier to review if it did not mix code movement with functional changes. I'd suggest having a commit where this code is moved to `shell.go` without any changes, and  then a commit porting the Windows code to use `detectShellByCheckingProcessTree` with these `crcos.IsPresentInList` and such.